### PR TITLE
Update Azure dynamic inventory authentication using user / pass

### PR DESCRIPTION
**Pull request by pmarques**
##### SUMMARY

If user password contains a '<' or '>' (greater or less then char) the dynamic inventory fails with this error:
```
, InvalidClientIdError: (invalid_request) AADSTS90013: Invalid input received from the user.
Trace ID: *****
Correlation ID: *********
Timestamp: 2018-01-21 19:39:44Z
```

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

contrib/inventory/azure_rm.py

##### ADDITIONAL INFORMATION

This update comes after this discussion:
 * https://github.com/Azure/azure-sdk-for-python/issues/1629#issuecomment-357558411

This patch applies recommendations from Microsoft documentation:
 * https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python#mgmt-auth-legacy

### CODE REVIEW

File: contrib/inventory/azure_rm.py [hunk 1]
```diff
@@ -311,10 +313,17 @@ def __init__(self, args):
...
+            LEGACY_CLIENT_ID = '04b07795-8ddb-461a-bbee-02f9e1bf7b46'
+
+            LOGIN_ENDPOINT = self._cloud_environment.endpoints.active_directory
+            RESOURCE = self._cloud_environment.endpoints.active_directory_resource_id
+            context = adal.AuthenticationContext(LOGIN_ENDPOINT + '/' + tenant)
+            self.azure_credentials = AdalAuthentication(context.acquire_token_with_username_password,
+                                                        RESOURCE,
+                                                        self.credentials['ad_user'],
+                                                        self.credentials['password'],
+                                                        LEGACY_CLIENT_ID)
```
Comment (yuwzho) [inline contrib/inventory/azure_rm.py]: What's the legacy_client_id, seems it is no need: https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python#mgmt-auth-token

**Review by yuwzho — CHANGES_REQUESTED** [review]

File: contrib/inventory/azure_rm.py (hunk 1 above)
Comment (pmarques) [reply contrib/inventory/azure_rm.py]: tbh, I din't find any good reference for that constant but that seems to be the CLI Client ID _backed_ in old Azure CLI [azure-xplat-cli/lib/util/constants.js#L47](https://github.com/Azure/azure-xplat-cli/blob/436834d678dee1facd2e3b1d7b2f5af930847b90/lib/util/constants.js#L47)

The example in that page unfortunately is only using **acquire_token_with_client_credentials** which is for Service Principal.

Comment (yuwzho) [reply contrib/inventory/azure_rm.py]: Please follow the docs.microsoft.com to use the Authentication

Comment (pmarques) [reply contrib/inventory/azure_rm.py]: @yuwzho which documentation? like I told you, I followed the doc  / SDK. As I said in the PR (on top) this was shared by a guy from Microsoft in their SDK repo.

@lmazuel Can you check if this follows your recommendations?

Comment (lmazuel) [reply contrib/inventory/azure_rm.py]: UserPasswordCredentials is supposed to be deprecated in favor of ServicePrincipalCredentials. However, we still provide it for backward compatibility.
The UserPassword flow requires a client_id. But usually people didn't have one (they just have their user/password). A very long time ago (I wasn't even at Microsoft yet), implementation was made to use by default the CLI 1.0 client_id, for people who are still using this flow. [This is what is done in current implementation in msrestazure as well](https://github.com/Azure/msrestazure-for-python/blob/ac8495ab2881e0dde9b65f165baa5dc6b5ec7e5b/msrestazure/azure_active_directory.py#L356-L357).

I would suggest you allow people to give their own client_id if using user/password, but to fallback to the one from CLI 1.0 if not provided (likely most common scenario)

TL;DR; Yes, use `self.credentials.get('client_id', LEGACY_CLIENT_ID)` in case of UserPassword and I'm fine.

**Review by lmazuel — APPROVED** [review]
LGTM

**Comment by yuwzho** [comment]
shipit

**Comment by pmarques** [comment]
@yuwzho I'm afraid this is stalled by your review, can you approve my changes?

**Review by yuwzho — APPROVED** [review]

**Comment by yuwzho** [comment]
@nitzmahone Could you please help to review and merge it?

**Comment by yuwzho** [comment]
bot_status

**Comment by yuwzho** [comment]
bot_broken

**Comment by zikalino** [comment]
shipit

**Comment by jborean93** [comment]
bot_status

File: contrib/inventory/azure_rm.py [hunk 2]
```diff
@@ -311,10 +313,18 @@ def __init__(self, args):
             tenant = self.credentials.get('tenant')
             if not tenant:
                 tenant = 'common'
-            self.azure_credentials = UserPassCredentials(self.credentials['ad_user'],
-                                                         self.credentials['password'],
-                                                         tenant=tenant,
-                                                         cloud_environment=self._cloud_environment)
+
+            # # Default to Xplat Client ID
+            LEGACY_CLIENT_ID = '04b07795-8ddb-461a-bbee-02f9e1bf7b46'
```
Comment (yungezz) [inline contrib/inventory/azure_rm.py:318]: put az cli client id in ansible code raises some concern since it's not right thing to do.

Comment (pmarques) [reply contrib/inventory/azure_rm.py:318]: @yungezz that was discussed in a review on the 22nd of Jan, https://github.com/ansible/ansible/pull/35156#discussion_r163302270

**Comment by Fred-sun** [comment]
@pmarques The PR exist conflict files, Could you help to resolve it? Thanks!

**Comment by pmarques** [comment]
I think PR #37909 resolved the issue with the password using the new authentication API. We can closed this issues
